[STEP 1] LJ, 준호

limjy0 opened the pull request:
안녕하세요 🍞브래드(@bradheo65)
LJ(@lj-7-77), 준호(@junho15) 입니다.

STEP 1
### 고민했던 점
- 사용자 입력값과 컴퓨터 임의의 패를 비교할 때, 비교조건 내용을 고민하였습니다.
> 예: 사용자 입력값이 가위일 경우
> 방법(1) 1. 컴퓨터의 패와 같은지?(True/False) -> 2. 컴퓨터의 패가 바위인가?(True/False)
 방법(2) 1. 컴퓨터의 패가 바위인가?(True/False) -> 2. 컴퓨터의 패가 보인가?(True/False)

➡️ 방법(2)의 경우 사용자 입력값에 따라 비교조건이 전부 달라집니다. 
(가위일땐 바위,보,  바위일땐 가위,보, 보일땐 바위,가위 )
방법(1)의 경우에는 첫번째 비교조건은 컴퓨터의 패와 같은것으로 통일 할 수 있습니다.
(가위일땐 동일여부,바위,  바위일땐 동일여부,보, 보일땐 동일여부,가위)
그래서 방법(1)로 구현하였습니다.

---
- RockPaperScissors, WinLoseDraw 열거형을 사용해서 선택지를 제한했습니다.
- RockPaperScissors에는 Int 타입의 원시 값을 추가해서 원시 값으로 항목을 찾을 수 있도록 했습니다.
- WinLoseDraw에는 결과 출력 메서드를 추가해서 승부 결과별로 다르게 출력해 줄 수 있게 했습니다.

---
- RockPaperScissorsGame 구조체에서 startGame를 제외한 나머지 메서드에는 접근 제어 키워드 private를 추가해서 구조체 밖에서 접근할 수 없도록 했습니다.
- 클래스의 상속 기능이 필요가 없고 연관된 기능의 집합을 단순히 캡슐화하는 것을 목적으로 하기 때문에 클래스 대신 구조체를 사용했습니다.
- RockPaperScissors, WinLoseDraw, RockPaperScissorsGame 파일로 분리해서 main에는 실행 코드만 남겨두었습니다.

---
- 코드의 가독성을 위해 Naming을 -Menu접미사와 -Hand접미사로 구분지었습니다.
- 사용자에게 직접 입력받은 값과 입력 받는 메서드의 이름은 selectedUserMenu, getSelectedUserMenu 와 같이 Menu접미사를 붙였습니다.
- RockPaperScissors 타입의 변수명은 computerHand, userHand 와 같이 Hand 접미사를 붙였습니다.
- 처음에는 가위, 바위, 보를 1, 2, 3의 Int 타입으로 표현했는 데, 가독성을 위해 RockPaperScissors 열거형을 사용했습니다.

---

### 조언을 얻고 싶은 부분
- RockPaperScissorsGame.swift 파일
~~~swift
    // 50~52줄
    private func getComputerHand() -> RockPaperScissors {
        return RockPaperScissors.allCases.randomElement() ?? .rock
    }
~~~
- getComputerHand 메서드 내부에서 랜덤 요소를 가져오는 메서드의 반환값이 `optional`타입이였습니다.
- 고민했던 방법
    1. getComputerHand 메서드를 호출할 때 반환값을 `optional`로 받아와서 guard let 또는 if let 으로 처리하는 방법
    2. nil병합연산자로 처리하는 방법 <- nil이 나올 가능성이 없기 때문에 이 방법으로 코드작성하였습니다.
- 이런 경우 옵셔널 처리를 어떻게 하는 게 좋을지 궁금합니다.

---

junho15 committed:
feat: 가위바위보 메뉴 출력 함수, 사용자 패 입력 함수 추가

limjy0 committed:
feat: 사용자의 입력값 검증 함수 추가

junho15 committed:
refactor: 함수 추가, 메서드 순서 수정, 반환타입 변경

- 임의의 컴퓨터패 만드는 함수 추가
- 호출 순서대로 메서드 순서 수정
- 메서드 반환타입 변경

limjy0 committed:
refactor: 열거형 추가, 반환타입과 메소드 리팩토링

- 가위바위보 열거형타입 추가
- 반환타입 변경
- 메소드 반환타입 변경
- 변수명, 메소드명 변경

junho15 committed:
feat: 승패 판단 함수 추가, 승무패 열거형 추가

limjy0 committed:
refactor: 결과출력메소드 추가 ,메소드 은닉화

- 기존 결과출력문을 열거형WinLoseDraw에 메소드로 추가하여 대체
- startGame메소드를 제외한 나머지 메소드를 은닉화

junho15 committed:
refactor: 파일 분리

- RockPaperScissors, WinLoseDraw, RockPaperScissorsGame 파일 추가
- main에는 실행문만 남기고 각 파일로 타입 이동

limjy0 committed:
refactor: startGame메소드 수정

- 사용자입력값을 열거형타입으로 변환해주는 기능구현 과정에서 nil병합 연산자 대신 if let문을 활용하여 optional 처리

junho15 committed:
docs: README 파일에 STEP1 내용 추가

bradheo65 commented on RockPaperScissors/RockPaperScissors/RockPaperScissorsGame.swift:
```diff
@@ -0,0 +1,75 @@
+
+struct RockPaperScissorsGame {
+    func startGame() {
+        let selectedUserMenu: Int = getSelectedUserMenu()
```
타입 명시를 해주신 부분도 있고 안해주신 부분도 있는 것 같아요!! 전반적으로 통일해 주시면 좋을 것 같습니다!!

jylim72 replied:
넵, 수정하였습니다.😃
```swift
let rockPaperScissorsGame: RockPaperScissorsGame = RockPaperScissorsGame()
```

bradheo65 commented on RockPaperScissors/RockPaperScissors/RockPaperScissorsGame.swift:
```diff
@@ -0,0 +1,75 @@
+        return selectedUserMenu
+    }
+    
+    private func printMenu() {
+        print("가위(1), 바위(2), 보(3)! <종료 : 0> : ", terminator: "")
+    }
```
`printMenu()` 네이밍이 좀 더 명확하게 어떠한 메뉴 인지 알 수 있게끔 네이밍이 되면 더 좋을 것 같아요!!!

jylim72 replied:
네! 사용자에게 메뉴를 출력하는 것이라서 printUserMenu로 변경하였습니다😃
```swift
    private func printUserMenu() {
        print("가위(1), 바위(2), 보(3)! <종료 : 0> : ", terminator: "")
    }
```

bradheo65 commented on RockPaperScissors/RockPaperScissors/RockPaperScissorsGame.swift:
```diff
@@ -0,0 +1,75 @@
+        }
+    }
+    
+    private func getSelectedUserMenu() -> Int {
+        printMenu()
+        guard let selectedUserMenu: Int = Int(readLine() ?? ""),
+              isCorrectUserMenu(selectedUserMenu) else {
+            print("잘못된 입력입니다. 다시 시도해주세요.")
+            return getSelectedUserMenu()
+        }
+        return selectedUserMenu
+    }
```
하나의 메소드에서 너무 많은 기능이 있는 것 같아요!! 
조금 더 가볍게 만들 수 있을 것 같습니다!!

bradheo65 replied:
> example

```swift
    private func getSelectedUserMenu() -> Int {
        printMenu()
        guard let userInput = userInput() else {
            return 0
        }
        guard isCorrectUserMenu(userInput) else {
            print("잘못된 입력입니다. 다시 시도해주세요.")
            return getSelectedUserMenu()
        }
        
        return userInput
    }
    
    private func printMenu() {
        print("가위(1), 바위(2), 보(3)! <종료 : 0> : ", terminator: "")
    }
    
    private func userInput() -> Int? {
        let selectedUserMenu: Int? = Int(readLine() ?? "")
        
        return selectedUserMenu
    }
```
이런식으로 변경하면 어떨까요~?

bradheo65 replied:
추가적으로 `getSelectedUserMenu()` 메소드에 if else를 통해 사용한다면 
```swift
    private func getSelectedUserMenu() -> Int {
        printMenu()
        guard let userInput = userInput() else {
            return 0
        }
        
        if isCorrectUserMenu(userInput) {
            return userInput
        } else {
            print("잘못된 입력입니다. 다시 시도해주세요.")
            return getSelectedUserMenu()
        }
    }
```
이런 형태가 될 수 있을 것 같아요!!
로직은 동일 하지만 코드를 보았을 때 가독성에 대해 차이가 있을 것 같아요! 두 분의 의견이 궁금합니다!

jylim72 replied:
아..! guard문 조건을 분리하는 방법이 있었네요!!👍
최대한 간결하게 작성하려다보니 가독성을 놓쳤습니다.
예시와 같이 사용자가 잘못입력했을 경우 조건들을 guard로 처리 후 값을 반환하는것으로 수정하였습니다. 감사합니다!!

bradheo65 commented on RockPaperScissors/RockPaperScissors/RockPaperScissorsGame.swift:
```diff
@@ -0,0 +1,75 @@
+    }
+    
+    private func getComputerHand() -> RockPaperScissors {
+        return RockPaperScissors.allCases.randomElement() ?? .rock
```
default 값을  `.rock`를 해주신 이유가 있을까요??
옵셔널 타입으로 반환을 하면 어떻게 될까요??🤔

jylim72 replied:
반환타입에 맞추려 임의의값을 넣었는데 해결방법이 있었네요!  옵셔널 타입으로 반환하고 함수로 호출하는쪽에서 옵셔널처리를 했습니다.🤓

bradheo65 commented on RockPaperScissors/RockPaperScissors/RockPaperScissorsGame.swift:
```diff
@@ -0,0 +1,75 @@
+    }
+    
+    private func compareUserHandWithComputerHand(userHand: RockPaperScissors, computerHand: RockPaperScissors) -> WinLoseDraw {
+        guard userHand != computerHand else {
+            return .draw
+        }
```
guard로 함수 시작 전에 미리 검사하고 가는 거 멋져요!!👍👍

bradheo65 commented on RockPaperScissors/RockPaperScissors/WinLoseDraw.swift:
```diff
@@ -0,0 +1,23 @@
+//  Created by LJ, 준호 on 2022/08/23.
+//
+
+enum WinLoseDraw {
```
해당 enum 타입에 있는 요소들에 대한 네이밍을 해주신 것 같아요!!
제 생각이지만 이 enum 타입은 게임결과?? 에 대한 네이밍이 어울릴 것 같은데 어떻게 생각하시나요?🤔

jylim72 replied:
저희도 동의합니다. startGame메서드에서 gameResult상수가 있어서 무의식적으로 enum네이밍과 동일하면 안된다고 생각했던 것 같아요. 조금 더 어울리는 gameResult로 변경하였습니다😃 
```swift
enum gameResult {
```

bradheo65 commented on RockPaperScissors/RockPaperScissors/WinLoseDraw.swift:
```diff
@@ -0,0 +1,23 @@
+        case .draw:
+            print("비겼습니다!")
+        }
+    }
```
연산 프로퍼티 사용 정말 좋습니다!! 👍👍
printResult 보단 result도 괜찮아보여요!! 
나중에 접근 시 `WinLoseDraw.printResult()` 보단 `WinLoseDraw.result()`가 더 어울린 것 같아요!! 
두 분의 의견이 궁금합니다!!

jylim72 replied:
나중에 접근할때의 코드까지는 생각을 못했네요🤩
gameResult열거형에서 printResult 메서드를 result프로퍼티로 변경하였습니다.😀
```swift
 var result: String {
        switch self {
        case .win:
            return "이겼습니다!"
        case .lose:
            return "졌습니다!"
        case .draw:
            return "비겼습니다!"
        }
    }
```

bradheo65 commented on RockPaperScissors/RockPaperScissors/main.swift:
```diff
@@ -1,10 +1,8 @@
-print("Hello, World!")
+//
 
+let rockPaperScissorsGame = RockPaperScissorsGame()
```
이 부분도 접근제한자 걸어주면 어떻게 될까요?

jylim72 replied:
아!! 전역변수에도 접근제한자를 사용할 수 있군요
접근제한자 사용여부에 대해 좀더 공부해보겠습니다.😃
```swift
private let rockPaperScissorsGame: RockPaperScissorsGame = RockPaperScissorsGame()
```

bradheo65 commented on RockPaperScissors/RockPaperScissors/RockPaperScissorsGame.swift:
```diff
@@ -0,0 +1,75 @@
+    func startGame() {
+        let selectedUserMenu: Int = getSelectedUserMenu()
+       
+        if let userHand: RockPaperScissors = RockPaperScissors.init(rawValue: selectedUserMenu) {
```
`if let`을 사용하신 특별한 이유가 있을까요!?

jylim72 replied:
`guard let`과 `if let` 모두 사용 가능하지만,
userHand를 if문 밖에서 사용할 필요가 없기때문에 `if let`을 사용하였습니다.🤔

bradheo65 commented on RockPaperScissors/RockPaperScissors/RockPaperScissorsGame.swift:
```diff
@@ -0,0 +1,75 @@
+            gameResult.printResult()
+            if gameResult == .draw {
+                startGame()
+            }
```
`startGame`메소드가 많은 일을 하는 것 같아요!! 메소드를 나눠주면 좋을 것 같습니다.
보통 하나의 메소드가 하나의 기능을 담당하는게 가독성이 좋아요! 유지보수성에서도 좋구요!!

jylim72 replied:
startGame메서드에서 컴퓨터의패(computerHand)를 확인하는 기능을 사용자의 패와 컴퓨터의 패를 비교하는 함수로 이동하였습니다.
1개 if문에서 optional binding처리하도록 코드 수정하였습니다.
```swift
if let userHand: RockPaperScissors = RockPaperScissors.init(rawValue: selectedUserMenu),
    let gameResult: gameResult = compareComputerHand(with: userHand) {
    print(gameResult.result)
    if gameResult == .draw {
        startGame()
    }
}
```

bradheo65 reviewed (commented):
안녕하세요!! LJ, 준호 이번 프로젝트 잘 부탁드립니다!! 🙇🏻‍♂️🙇🏻‍♂️
STEP1 구현하시느라 정말 고생많으셨습니다.
제가 했었을 때와 수준이 완전 다르네요.. 완전 멋집니다. 👍👍
타입별로 파일을 분리해주신 것도 정말 좋습니다!
그리고 제가 리뷰한 코드들은 정답이 아니고 참고용으로만 생각해주세요!! 두 분이 왜 구현했고, 왜 필요한 지에 대한 생각이 제일 중요 합니다!

## 고민했던 점
고민했던 점을 읽어보니 코멘트를 남기면 좋은 부분이 있어 남겨드립니다. 
- 클래스의 상속 기능이 필요가 없고 연관된 기능의 집합을 단순히 캡슐화하는 것을 목적으로 하기 때문에 클래스 대신 구조체를 사용했습니다.

프로젝트를 사용함에 있어 클래스와 구조체를 사용을 많이 하실텐데 이렇게 명확한 이유가 있다면 class와 struct 중 어떠한 것을 사용해도 저도 괜찮다고 생각합니다!! 특별히 기능상에 문제가 없다면요!! 
워낙 장단점이 있는 타입이라,, 고민한 점 최곱니다!!👍
추가적으로 [공식문서-class and structe](https://docs.swift.org/swift-book/LanguageGuide/ClassesAndStructures.html#) 도 한번 보시면 도움이 될 것 같습니다!
## 조언을 얻고 싶은 부분
### Q1. RockPaperScissorsGame.swift 파일
```swift
    // 50~52줄
    private func getComputerHand() -> RockPaperScissors {
        return RockPaperScissors.allCases.randomElement() ?? .rock
    }
```
getComputerHand 메서드 내부에서 랜덤 요소를 가져오는 메서드의 반환값이 optional타입이였습니다.
고민했던 방법
getComputerHand 메서드를 호출할 때 반환값을 optional로 받아와서 guard let 또는 if let 으로 처리하는 방법
nil병합연산자로 처리하는 방법 <- nil이 나올 가능성이 없기 때문에 이 방법으로 코드작성하였습니다.
이런 경우 옵셔널 처리를 어떻게 하는 게 좋을지 궁금합니다.

## 답변
### A1. 반환 값 타입을 옵셔널로 하게 되면 어떻게 될까요??

> example
```swift
    private func getComputerHand() -> RockPaperScissors? {
```
이런 식으로 옵셔널 값으로 넘겨주면 옵셔널 처리를 따로 안해줘도 될 것 같아요!! 
하지만 이렇게 되면 해당 메소드를 사용하는 곳에서 옵셔널 처리를 해줘야 되겠죠!

limjy0 committed:
refactor: 리뷰 수정사항 반영

- WinLoseDraw열거형 이름을 gameResult로 변경
- gameResult열거형에서 printResult메서드를 result프로퍼티로 변경
- rockPaperScissorsGame 상수선언에 타입 명시
- rockPaperScissorsGame 상수에 접근제한자 private 추가
- startGame메서드에서 computerHand을 1개 if문에서 optional binding처리하도록 코드 수정
- printMenu메서드 이름을 printUserMenu로 변경
- getComputerHand메서드 반환타입을 optional로 변경
- compareUserHandWithComputerHand메서드 이름을 compareComputerHand로 변경
- compareComputerHand메서드 매개변수를 하나만 받도록 수정
- compareComputerHand메서드 내부에서 computerHand를 생성하도록 수정
- compareComputerHand메서드 반환타입을 optional로 변경

junho15 committed:
refactor: userInput 메서드 추가하여 getSelecteduserMenu 메서드의 기능 분리

limjy0 commented:
STEP 1 리뷰 고생하셨습니다.👏👏👏

junho15 committed:
refactor: gameResult 타입 이름을 대문자로 시작하도록 GameResult로 수정

limjy0 commented:
gameResult 열거형 이름 GameResult로 수정하였습니다.
merge 해주시면 될것같아요 감사합니다👍👍👍
